velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp

```diff
@@ -0,0 +1,353 @@
+      uint64_t offset,
+      const std::vector<folly::Range<char*>>& buffers) const override {
+    // 'buffers' contains Ranges(data, size)  with some gaps (data = nullptr) in
+    // between. This call must populate the ranges (except gap ranges)
+    // sequentially starting from 'offset'. If a range pointer is nullptr, the
+    // data from stream of size range.size() will be skipped.
+    size_t length = 0;
+    for (const auto range : buffers) {
+      length += range.size();
+    }
+    // TODO: allocate from a memory pool
+    std::string result(length, 0);
+    preadInternal(offset, length, static_cast<char*>(result.data()));
+    size_t resultOffset = 0;
+    for (auto range : buffers) {
+      if (range.data()) {
+        memcpy(range.data(), &(result.data()[resultOffset]), range.size());
+      }
+      resultOffset += range.size();
+    }
+    return length;
+  }
+
+  uint64_t size() const override {
+    return length_;
+  }
+
+  uint64_t memoryUsage() const override {
+    // TODO: Check if any buffers are being used by the GCS library
+    return sizeof(gcs::Client) + kUploadBufferSize + 2 * sizeof(std::string) +
+        sizeof(int64_t);
+  }
+
+  bool shouldCoalesce() const final {
+    return false;
+  }
+
+  std::string getName() const override {
+    return key_;
+  }
+
+  uint64_t getNaturalReadSize() const override {
+    return kUploadBufferSize;
+  }
+
+ private:
+  // The assumption here is that "position" has space for at least "length"
+  // bytes.
+  void preadInternal(uint64_t offset, uint64_t length, char* position) const {
+    gcs::ObjectReadStream stream = client_->ReadObject(
+        bucket_, key_, gcs::ReadRange(offset, offset + length));
+    VELOX_CHECK_GCS_OUTCOME(
+        stream.status(), "Failed to get GCS object", bucket_, key_);
+
+    stream.read(position, length);
+    VELOX_CHECK_GCS_OUTCOME(
+        stream.status(), "Failed to get read object", bucket_, key_);
+    bytesRead_ += length;
+  }
+
+  gcs::Client* client_;
+  std::string bucket_;
+  std::string key_;
+  std::atomic<int64_t> length_ = -1;
+};
+
+} // namespace
+
+namespace filesystems {
+
+class GCSConfig {
+ private:
+  const Config* FOLLY_NONNULL config_;
+  std::shared_ptr<gc::Credentials> credentials_;
+
+ public:
+  GCSConfig(const Config* config) : config_(config) {
+    std::string cred = config_->get("hive.gcs.credentials", std::string(""));
+    // Cred expected in json format, change it to something else?
+    if (!cred.empty()) {
+      credentials_ = gc::MakeServiceAccountCredentials(cred);
+    }
+  }
+
+  std::string endpointOverride() const {
+    return config_->get("hive.gcs.endpoint", std::string(""));
+  }
+
+  std::string scheme() const {
+    return config_->get("hive.gcs.scheme", std::string("https"));
+  }
+
+  std::shared_ptr<gc::Credentials> credentials() const {
+    return credentials_;
+  }
+};
+
+class GCSFileSystem::Impl {
+ public:
+  Impl(const Config* config) : gcsConfig_(config) {}
+
+  ~Impl() {}
```

Comment:
Any reason for this destructor over default ?

Reply:
Correct, I will use the default destructor.